src/PiggyAuth/Tasks/AsyncLoginTask.php

```diff
@@ -0,0 +1,129 @@
+        $this->potentialPassword = $potentialPassword;
+        $this->originAuth = $originAuth;
+        $this->mode = $mode;
+        $this->cost = $cost;
+    }
+
+    public function onRun()
+    {
+        if (strpos($this->originAuth, "ServerAuth") !== false) {
+            $auth = explode("_", $this->originAuth);
+            if (isset($auth[0]) && isset($auth[1])) {
+                if (hash($auth[1], $this->potentialPassword) == $this->passwordHash) {
+                    $this->setResult(['authenticated' => true, 'updatePlayer' => true, 'rehashedPassword' => $this->needsRehashPassword()]);
+                    return;
+                }
+                $this->setResult(['authenticated' => false, 'updatePlayer' => false]);
+                return;
+            }
+        }
+
+        switch ($this->originAuth) {
+            case "SimpleAuth":
+                if (hash_equals($this->passwordHash, $this->hashSimpleAuth(strtolower($this->playerName), $this->potentialPassword))) {
+                    $this->setResult(['authenticated' => true, 'updatePlayer' => true, 'rehashedPassword' => $this->needsRehashPassword()]);
+                    return;
+                }
+                $this->setResult(['authenticated' => false, 'updatePlayer' => false]);
+                return;
+            case "PiggyAuth":
+            default:
+                if (password_verify($this->potentialPassword, $this->passwordHash)) {
+                    $this->setResult(['authenticated' => true, 'updatePlayer' => false, 'rehashedPassword' => $this->needsRehashPassword()]);
+                    return;
+                }
+                $this->setResult(['authenticated' => false, 'updatePlayer' => false]);
+                return;
+        }
+    }
+
+    public function hashSimpleAuth($salt, $password)
+    {
+        return bin2hex(hash("sha512", $password . $salt, true) ^ hash("whirlpool", $salt . $password, true));
+    }
+
+    public function needsRehashPassword()
+    {
+        $options = ['cost' => $this->cost];
+        if (password_needs_rehash($this->passwordHash, PASSWORD_BCRYPT, $options)) {
+            return password_hash($this->potentialPassword, PASSWORD_BCRYPT, $options);
+        }
+        return null;
+    }
+
+    public function onCompletion(Server $server)
+    {
+        $plugin = $server->getPluginManager()->getPlugin('PiggyAuth');
+
+        if($plugin->isDisabled() || !(($player = $server->getPlayerExact($this->playerName)) instanceof Player)){
+            return;
+        }
+
+        if($this->getResult()['updatePlayer']){
+            $plugin->sessionmanager->getSession($player)->updatePlayer("auth", "PiggyAuth");
+        }
+
+        if (!$this->getResult()['authenticated']) { // Failed
+            if ($plugin->getConfig()->getNested("key.enabled")) {
+                if ($this->potentialPassword === $plugin->getKey($this->potentialPassword)) { // Please look at this @MCPEPIG
```

Comment:
This looks fine except the parameter for getKey() has to be the owner's password (owner's username is provided in config)